### ISSUE TITLE
Legion/GasNET: not identical behavior on multiple nodes

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#776]
I have a CFD solver based on Legion. I am testing multi-node capability using GasNET on NASA Pleiades supercomputer. GasNET is built in release.

#### Issue
I don't get the same behavior going from 1 node to 2 nodes:
- 1 node with GasNET linked to Legion is identical to Legion without GasNET
- 2 nodes with GasNET linked to Legion produces different results

The solutions are compared by computing an error value at the end of the simulation and thus does not involve dumping the solution into HDF5.

**Comment by lightsighter** [comment]
Please compile the runtime with detailed Legion Spy support and then produce a Legion Spy trace for the two node case that reproduces the issue. If you need directions on how to do this please tell me which build system you are using.

**Comment by bandokihiro** [comment]
following Wonchan's recommendation, getting rid of reduction operations solved the issue

**Comment by bandokihiro** [comment]
I am using the cmake build system

**Comment by lightsighter** [comment]
So we think that non-deterministic ordering of reduction applications is responsible for the difference? That kind of error would only show up in the least significant bits of data due to floating point round-off error. Is that consistent with what you were seeing?

**Comment by bandokihiro** [comment]
no the difference cannot be explained by round-off errors

**Comment by lightsighter** [comment]
Ok, then I would still like to see a Legion Spy trace. Add this to your cmake build line:
`cmake -DCMAKE_CXX_FLAGS="-DLEGION_SPY" ...` and run with `-level legion_spy=2 -logfile log_%.spy` in your command line. We need to run as small a problem (both spatially and temporally) as possible in order to reproduce the problem. Please upload the log files from a run that produces the "wrong answer" here.

**Comment by bandokihiro** [comment]
so I am afraid this issue cannot be reproduced in debug

**Comment by lightsighter** [comment]
I didn't say to use debug, you only need legion spy support.

**Comment by bandokihiro** [comment]
In the past I was told to use debug when using legion spy. Let me try again.

**Comment by bandokihiro** [comment]
[legion_spy_failed_run.zip](https://github.com/StanfordLegion/legion/files/4351992/legion_spy_failed_run.zip)
This was the smallest I could do that failed

**Comment by magnatelee** [comment]
I just had a private conversation with @bandokihiro and it looks like there is a bug in his reduction operator; his non-exclusive implementation is not thread safe, which could lead to all sorts of correctness issues. I think that we should fix that issue first and then check if this and issue #770 still persist.

**Comment by lightsighter** [comment]
> In the past I was told to use debug when using legion spy. Let me try again.

That was true up to about two years ago. It is no longer true anymore.

**Comment by bandokihiro** [comment]
This issue was fixed with a correct reduction implementattion. Thanks.
